[JSC] Don't use destructuring assignment in builtin `PromiseConstructor.js`

sosukesuzuki opened the pull request:
#### 66909c070dee425617357cb6c5c1af433eb9ae8c
<pre>
[JSC] Don&apos;t use destructuring assignment in builtin `PromiseConstructor.js`
<a href="https://bugs.webkit.org/show_bug.cgi?id=278958">https://bugs.webkit.org/show_bug.cgi?id=278958</a>

Reviewed by Yusuke Suzuki.

This patch removes destructuring assignment from builtin `PromiseConstructor.js`.

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(allSettled):

Canonical link: <a href="https://commits.webkit.org/283040@main">https://commits.webkit.org/283040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b225590339b5fb94b8bef1400679173ffded6959

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64989 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17603 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69013 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15595 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52139 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52209 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10767 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56245 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32831 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37681 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14471 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58102 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59593 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70718 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64232 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59538 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56305 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59769 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7375 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1050 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85998 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9855 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40168 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15158 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41245 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42426 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40989 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/PromiseConstructor.js:
```diff
@@ -246,9 +246,9 @@ function allSettled(iterable)
             @putByValDirect(values, index, @undefined);
             var nextPromise = promiseResolve.@call(this, value);
-            var [resolveElement, rejectElement] = newResolveRejectElements(index);
+            var resolveRejectElements = newResolveRejectElements(index);
```
Let's return object instead of array too.

sosukesuzuki replied:
Does your comment mean the following code?

```javascript
var resolveRejectElements = newResolveRejectElements(index);
++remainingElementsCount;
nextPromise.then(resolveRejectElements.resolve, resolveRejectElements.reject);
```

If so, I believe we should keep it as an array because this code would cause tests like [this one](https://github.com/tc39/test262/blob/main/test/built-ins/Promise/all/resolve-element-function-name.js) to fail.

Constellation replied:
You can avoid that issue like this.
```
var object = {
    test: (void 0, () = > { })
};
object.test.name // ""
```

We should change that since the above object literal based one is much better in terms of efficacy compared to array.

Constellation replied:
But I think another option you can take is just inline entire function, and using let scoped variables instead, probably that's better in terms of efficacy.
```
let currentIndex = index++;
let alreadyCalled = false;
nextPromise.then(
            (value) => {
                if (alreadyCalled)
                    return @undefined;
                alreadyCalled = true;
                var obj = {
                    status: "fulfilled",
                    value
                };
                @putByValDirect(values, currentIndex, obj);
                --remainingElementsCount;
                if (remainingElementsCount === 0)
                    return promiseCapability.resolve.@call(@undefined, values);
                return @undefined;
            },
            (reason) => {
                if (alreadyCalled)
                    return @undefined;
                alreadyCalled = true;
                var obj = {
                    status: "rejected",
                    reason
                };
                @putByValDirect(values, currentIndex, obj);
                --remainingElementsCount;
                if (remainingElementsCount === 0)
                    return promiseCapability.resolve.@call(@undefined, values);
                return @undefined;
            });

sosukesuzuki replied:
Thank you. Your second option sounds good to me, I've updated the patch.

sosukesuzuki committed:
[JSC] Don't use destructuring assignment in builtin `PromiseConstructor.js`
https://bugs.webkit.org/show_bug.cgi?id=278958

Reviewed by Yusuke Suzuki.

This patch removes destructuring assignment from builtin `PromiseConstructor.js`.

* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(allSettled):

Canonical link: https://commits.webkit.org/283040@main
